query/src/main/scala/filodb/query/exec/ApplySubqueryRangeFunction.scala

```diff
@@ -0,0 +1,155 @@
+package filodb.query.exec
+
+import monix.reactive.Observable
+
+import filodb.core.query._
+import filodb.query.exec.rangefn.RangeFunction
+import filodb.query.exec.rangefn.RangeFunction.RangeFunctionGenerator
+import filodb.query.util.IndexedArrayQueue
+
+// sum_over_time(foo[6:3])[4:2], where now=18
+final case class ApplySubqueryRangeFunction(
+  subqueryExpRangeParams: RangeParams, // start = now-4, end=now, step = 2
+  underlyingLookbackMs: Long, // 6
+  underlyingStepMs: Long, // 3
+  subqueryRangeFunction: InternalRangeFunction
+) extends RangeVectorTransformer {
+  override def funcParams: Seq[FuncArgs] = Nil
+
+  override protected[exec] def args: String =
+    s"subqueryEprRangeParams=$subqueryExpRangeParams," +
+    s"subqueryRangeFunction=$subqueryRangeFunction," +
+    s"underlyingLookbackMs=$underlyingLookbackMs,underlyingStepMs=$underlyingStepMs"
+
+  def apply(
+    source: Observable[RangeVector],
+    querySession: QuerySession,
+    limit: Int,
+    sourceSchema: ResultSchema,
+    paramsResponse: Seq[Observable[ScalarRangeVector]]
+  ): Observable[RangeVector] = {
+
+    //val underlyingRangeStepMs = underlyingPeriodicSeriesRange.stepSecs * 1000
+
+    // Need to skip these many samples for window operations
+    //val skip = (subqueryStepMs / underlyingRangeStepMs).toInt
+    val valColType = ResultSchema.valueColumnType(sourceSchema)
+    val rangeFuncGen = RangeFunction.generatorFor(
+      sourceSchema,
+      Some(subqueryRangeFunction),
+      valColType,
+      querySession.queryConfig, funcParams, false
+    )
+    source.map { rv =>
+      IteratorBackedRangeVector(
+        rv.key,
+        new SubquerySlidingWindowIterator(
+          subqueryExpRangeParams.startSecs * 1000,
+          subqueryExpRangeParams.endSecs * 1000,
+          subqueryExpRangeParams.stepSecs * 1000,
+          underlyingLookbackMs,
+          underlyingStepMs,
+          rv.rows,
+          rangeFuncGen,
+          querySession.queryConfig
+        ),
+        None
+      )
+    }
+  }
+
+  override def schema(source: ResultSchema): ResultSchema = {
+    // since range function is NOT optional, the resulting vectors length will be
+    // of size 1, not the length of the underlying PeriodicSeries
+    source.copy(fixedVectorLen = Some(1))
+  }
+
+}
+
+// sum_over_time(foo[5m:1m])[3m:90s]
+class SubquerySlidingWindowIterator(
```

Comment:
Deserves a detailed scaladoc if we plan to maintain it. The functionality deviates from PromQL so the "expected outcome" needs to be documented.

Reply:
noted

Reply:
I put extra explanation but I will also make a task to follow up on the documentation.

Reply:
removed the entire class